flowmachine/flowmachine/core/subscriber_subset.py

```diff
@@ -0,0 +1,186 @@
+    """
+    Represents a subset given by a flowmachine query.
+    """
+
+    is_proper_subset = True
+
+    def __init__(self, flowmachine_query):
+        assert isinstance(flowmachine_query, Query)
+
+        self.subset_query = text(flowmachine_query.get_query()).columns(
+            column("subscriber")
+        )
+
+    def _make_query(self):
+        # Return a dummy string representing this subset. This is only needed
+        # because SubscriberSubsetBase currently inherits from Query, but will
+        # eventually be removed.
+        return "<SubsetFromFlowmachineQuery>"
+
+    def apply_subset(self, sql, *, subscriber_identifier=None):
+        """
+        Return a modified version of the input SQL query which has the subset applied.
+
+        Parameters
+        ----------
+        sql : sqlalchemy.sql.ClauseElement
+            The SQL query to which the subset should be applied.
+
+        subscriber_identifier : str
+            This argument is ignored for subsets of type 'SubsetFromFlowmachineQuery'.
+
+        Returns
+        ----------
+        sqlalchemy.sql.ClauseElement
+        """
+        assert isinstance(sql, ClauseElement)
+
+        tbl = sql.alias("tbl")
+
+        try:
+            subset = self.subset_query.distinct().alias("subset")
+        except AttributeError:
+            # This can happen if `self.subset_query` is a textual query
+            subset = self.subset_query.alias("subset")
+
+        res = select(tbl.columns).select_from(
+            tbl.join(subset, tbl.c.subscriber == subset.c.subscriber)
+        )
+
+        return res
+
+
+class ExplicitSubset(SubscriberSubsetBase):
+    """
+    Represents a subset given by an explicit list of subscribers.
+    """
+
+    is_proper_subset = True
+
+    def __init__(self, subscribers):
+        assert isinstance(subscribers, (list, tuple))
+        self.subscribers = subscribers
+
+    def _make_query(self):
+        # Return a dummy string representing this subset. This is only needed
+        # because SubscriberSubsetBase currently inherits from Query, but will
+        # eventually be removed.
+        return "<ExplicitSubset>"
+
+    def apply_subset(self, sql, *, subscriber_identifier):
+        """
+        Return a modified version of the input SQL query which has the subset applied.
+
+        Parameters
+        ----------
+        sql : sqlalchemy.sql.ClauseElement
+            The SQL query to which the subset should be applied.
+
+        subscriber_identifier : str
+            The column in the parent table which contains the subscriber information.
+
+        Returns
+        ----------
+        sqlalchemy.sql.ClauseElement
+        """
+        assert isinstance(sql, ClauseElement)
+        assert len(sql.froms) == 1
+        parent_table = sql.froms[0]
+        return sql.where(parent_table.c[subscriber_identifier].in_(self.subscribers))
+
+
+def make_subscriber_subset(subset):
```

Comment:
I find this name confusing. I would go for something like `make_subscriber_subset_factory` or `make_subscriber_subset_generator`

Reply:
Agreed it is quite confusing, especially because we have multiple different types of subsets at the moment (which should be cleaner again once the refactoring is complete). I'd like to avoid the word "factory" if possible because it sounds like tech lingo and some people (not me) have an aversion against it. Also, we're not really making a subscriber subset factory (rather, this _is_ the subscriber subset factory).

How about `make_subscriber_subset_object` or similar? Does this make it clearer that we're creating an object from "raw" input (such as a string or a list)?

Reply:
I actually think it's worth making it explicit in the names that this object plays an "active" role rather than just representing a "passive" subset. I'll rename the classes from "...Subset" to "...Subsetter" and call this function `make_subscriber_subsetter`. Hopefully this will make things more obvious.